### PR TITLE
[Telink] Enable -Wundef by default on Zephyr platform

### PR DESCRIPTION
**Pull request by s07641069**
Part PR of https://github.com/project-chip/connectedhomeip/pull/29582
Fix issue https://github.com/project-chip/connectedhomeip/issues/29593

### CODE REVIEW

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp [hunk 1]
```diff
@@ -1710,7 +1710,7 @@ CHIP_ERROR GenericThreadStackManagerImpl_OpenThread<ImplClass>::DoInit(otInstanc
     VerifyOrExit(otErr == OT_ERROR_NONE, err = MapOpenThreadError(otErr));
 
     // Enable automatic assignment of Thread advertised addresses.
-#if OPENTHREAD_CONFIG_IP6_SLAAC_ENABLE
+#if defined(OPENTHREAD_CONFIG_IP6_SLAAC_ENABLE)
```
Comment (Damian-Nordic) [inline src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp:1713]:
```suggestion
#if defined(OPENTHREAD_CONFIG_IP6_SLAAC_ENABLE) && OPENTHREAD_CONFIG_IP6_SLAAC_ENABLE
```
This can be set to 0 as well.

File: src/platform/Zephyr/ConfigurationManagerImpl.cpp [hunk 2]
```diff
@@ -206,7 +206,7 @@ void ConfigurationManagerImpl::DoFactoryReset(intptr_t arg)
 
 CHIP_ERROR ConfigurationManagerImpl::GetPrimaryWiFiMACAddress(uint8_t * buf)
 {
-#if CHIP_DEVICE_CONFIG_ENABLE_WIFI
+#if defined(CHIP_DEVICE_CONFIG_ENABLE_WIFI)
```
Comment (Damian-Nordic) [inline src/platform/Zephyr/ConfigurationManagerImpl.cpp:209]: All those, `CHIP_DEVICE_CONFIG_XXX` could be set to 0 as well, so checking they are defined only is error-prone. Alternatively, we should just make sure in `CHIPDevicePlatformConfig.h` that these are always defined and set to either 1 or 0.

**Review by Damian-Nordic — CHANGES_REQUESTED** [review]
Two comments for now, but in general, CHIP configs can be defined to 0 in the project config, so checking if they are defined is not the right solution. I would recommend making sure in `CHIPDevicePlatformConfig.h` that those configs are always defined.

**Review by Damian-Nordic — APPROVED** [review]

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp (hunk 1 above)
Comment (s07641069) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.hpp]: fixed

File: src/platform/Zephyr/ConfigurationManagerImpl.cpp (hunk 2 above)
Comment (s07641069) [reply src/platform/Zephyr/ConfigurationManagerImpl.cpp]: fixed

File: src/platform/nrfconnect/wifi/WiFiManager.h [hunk 3]
```diff
@@ -178,7 +178,7 @@ class WiFiManager
     static_assert(kConnectionRecoveryMinIntervalMs < kConnectionRecoveryMaxIntervalMs);
     static_assert(kConnectionRecoveryJitterMs <= kConnectionRecoveryMaxIntervalMs);
 
-#if CHIP_DEVICE_CONFIG_ENABLE_SED
+#if defined(CHIP_DEVICE_CONFIG_ENABLE_SED) && CHIP_DEVICE_CONFIG_ENABLE_SED
```
Comment (bzbarsky-apple) [inline src/platform/nrfconnect/wifi/WiFiManager.h:181]: I would expect this to be defined in some relevant config header (probably `platform/CHIPDeviceConfig.h`, see below).  So:

1.  Is that being done?  If not, it should be. 
2. Is this file including that config header?

Looking at where this macro is defined, we have `src/platform/Zephyr/CHIPDevicePlatformConfig.h` and the  `CHIPProjectConfig.h` of a bunch of examples.  Looks like all the examples in question set `chip_device_project_config_include = "<CHIPProjectConfig.h>"` so all of those defines would in fact land in `src/include/platform/CHIPDeviceConfig.h`.  If that header is then not defining a default, that should be fixed, and this file should include that header.

File: src/system/SystemConfig.h [hunk 4]
```diff
@@ -733,7 +733,7 @@ struct LwIPEvent;
  *  Defaults to enabled on Zephyr platforms that do not enable Zephyr POSIX layer.
  */
 #ifndef CHIP_SYSTEM_CONFIG_USE_ZEPHYR_SOCKETS
-#if CHIP_SYSTEM_CONFIG_USE_SOCKETS && defined(__ZEPHYR__) && CONFIG_NET_SOCKETS_POSIX_NAMES
+#if CHIP_SYSTEM_CONFIG_USE_SOCKETS && defined(__ZEPHYR__) && defined(CONFIG_NET_SOCKETS_POSIX_NAMES)
```
Comment (bzbarsky-apple) [inline src/system/SystemConfig.h:736]: Again, is the contract that prj.conf settings end up as defined-or-not, without boolean values?

File: src/system/SystemConfig.h [hunk 5]
```diff
@@ -765,7 +765,7 @@ struct LwIPEvent;
  *  Defaults to enabled on Zephyr platforms that enable CONFIG_EVENTFD.
  */
 #ifndef CHIP_SYSTEM_CONFIG_USE_ZEPHYR_EVENTFD
-#if defined(__ZEPHYR__) && CONFIG_EVENTFD
+#if defined(__ZEPHYR__) && defined(CONFIG_EVENTFD)
```
Comment (bzbarsky-apple) [inline src/system/SystemConfig.h:768]: CONFIG_EVENTFT is a Kconfig setting, right?  See questions about Kconfig contract above.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

**Comment by bzbarsky-apple** [comment]
@s07641069 And thank you for working on this; this is wonderful to see.  Sorry about all the nitpicks above; it's just that the parts I _do_ understand had some changes that were not quite right, and I am trying to make sure I understand the parts I don't understand yet.

File: src/platform/Zephyr/PlatformManagerImpl.cpp [hunk 6]
```diff
@@ -21,7 +21,7 @@
  *          for Zephyr platforms.
  */
 
-#if !CONFIG_NORDIC_SECURITY_BACKEND
+#if !defined(CONFIG_NORDIC_SECURITY_BACKEND)
```
[the remark this answers is not in the document]
Comment (Damian-Nordic) [reply src/platform/Zephyr/PlatformManagerImpl.cpp:24]: This is correct. Boolean Kconfig macros are either defined as true or not defined at all.

File: src/platform/Zephyr/ConfigurationManagerImpl.cpp [hunk 7]
```diff
@@ -206,7 +206,7 @@ void ConfigurationManagerImpl::DoFactoryReset(intptr_t arg)
 
 CHIP_ERROR ConfigurationManagerImpl::GetPrimaryWiFiMACAddress(uint8_t * buf)
 {
-#if CHIP_DEVICE_CONFIG_ENABLE_WIFI
+#if defined(CHIP_DEVICE_CONFIG_ENABLE_WIFI) && CHIP_DEVICE_CONFIG_ENABLE_WIFI
```
[the remark this answers is not in the document]
Comment (s07641069) [reply src/platform/Zephyr/ConfigurationManagerImpl.cpp]: Reworked:
defined CONFIG_WIFI_NRF700X which is Boolean Kconfig macros is defined as true or not defined at all:
`#ifndef CONFIG_WIFI_NRF700X`
`#define CONFIG_WIFI_NRF700X 0`
`#endif`

File: src/platform/Zephyr/DiagnosticDataProviderImpl.cpp [hunk 8]
```diff
@@ -39,7 +39,7 @@
 
 #include <malloc.h>
 
-#if CHIP_DEVICE_CONFIG_HEAP_STATISTICS_MALLINFO
+#if defined(CHIP_DEVICE_CONFIG_HEAP_STATISTICS_MALLINFO) && CHIP_DEVICE_CONFIG_HEAP_STATISTICS_MALLINFO
```
[the remark this answers is not in the document]
Comment (s07641069) [reply src/platform/Zephyr/DiagnosticDataProviderImpl.cpp]: Reworked:
added `#define CHIP_DEVICE_CONFIG_HEAP_STATISTICS_MALLINFO 0`

File: src/platform/Zephyr/ThreadStackManagerImpl.h [hunk 9]
```diff
@@ -29,7 +29,7 @@
 #include <zephyr/net/openthread.h>
 
 #include <openthread/thread.h>
-#if !CONFIG_SOC_SERIES_RISCV_TELINK_B9X
+#if !defined(CONFIG_SOC_SERIES_RISCV_TELINK_B9X)
```
[the remark this answers is not in the document]
Comment (s07641069) [reply src/platform/Zephyr/ThreadStackManagerImpl.h:32]: Boolean Kconfig macros is defined as true or not defined at all.

File: src/platform/nrfconnect/CHIPDevicePlatformConfig.h [hunk 10]
```diff
@@ -87,19 +87,25 @@
 #define CHIP_DEVICE_CONFIG_DEVICE_SOFTWARE_VERSION_STRING CONFIG_CHIP_DEVICE_SOFTWARE_VERSION_STRING
 #endif
 
+#if defined(CONFIG_NET_L2_OPENTHREAD) && CONFIG_NET_L2_OPENTHREAD
```
[the remark this answers is not in the document]
Comment (s07641069) [reply src/platform/nrfconnect/CHIPDevicePlatformConfig.h]: Reworked:
defined CONFIG_NET_L2_OPENTHREADwhich is Boolean Kconfig macros is defined as true or not defined at all:
```
#ifndef CONFIG_NET_L2_OPENTHREAD
#define CONFIG_NET_L2_OPENTHREAD 0
#endif
```

File: src/platform/nrfconnect/CHIPDevicePlatformConfig.h [hunk 11]
```diff
@@ -87,19 +87,25 @@
...
 #define CHIP_DEVICE_CONFIG_ENABLE_THREAD CONFIG_NET_L2_OPENTHREAD
+#endif
 
+#ifdef CONFIG_WIFI_NRF700X
 #define CHIP_DEVICE_CONFIG_ENABLE_WIFI CONFIG_WIFI_NRF700X
+#endif
 
-#if CHIP_DEVICE_CONFIG_ENABLE_WIFI
+#if defined(CHIP_DEVICE_CONFIG_ENABLE_WIFI) && CHIP_DEVICE_CONFIG_ENABLE_WIFI
```
[the remark this answers is not in the document]
Comment (s07641069) [reply src/platform/nrfconnect/CHIPDevicePlatformConfig.h]: Reworked:
defined CONFIG_WIFI_NRF700X which is Boolean Kconfig macros is defined as true or not defined at all:
```
#ifndef CONFIG_WIFI_NRF700X
#define CONFIG_WIFI_NRF700X 0
#endif
```

File: src/platform/nrfconnect/CHIPDevicePlatformConfig.h [hunk 12]
```diff
@@ -87,19 +87,25 @@
...
 #define CHIP_DEVICE_CONFIG_ENABLE_WIFI_STATION 1
 #define CHIP_DEVICE_CONFIG_ENABLE_WIFI_AP 0
 #else
 #define CHIP_DEVICE_CONFIG_ENABLE_WIFI_STATION 0
 #define CHIP_DEVICE_CONFIG_ENABLE_WIFI_AP 0
 #endif
 
+#ifdef CONFIG_BT
```
[the remark this answers is not in the document]
Comment (s07641069) [reply src/platform/nrfconnect/CHIPDevicePlatformConfig.h]: Reworked:
defined CONFIG_BT which is Boolean Kconfig macros is defined as true or not defined at all:
```
#ifndef CONFIG_BT
#define CONFIG_BT 0
#endif
```

File: src/platform/nrfconnect/CHIPDevicePlatformConfig.h [hunk 13]
```diff
@@ -217,7 +223,8 @@
 #endif // CONFIG_CHIP_ENABLE_DNS_CLIENT
 #endif // CONFIG_CHIP_ENABLE_DNSSD_SRP
 
-#ifdef CONFIG_CHIP_ENABLE_ICD_SUPPORT
+#if defined(CONFIG_CHIP_ENABLE_ICD_SUPPORT) && CONFIG_CHIP_ENABLE_ICD_SUPPORT && defined(CONFIG_CHIP_THREAD_SSED) &&               \
```
[the remark this answers is not in the document]
Comment (s07641069) [reply src/platform/nrfconnect/CHIPDevicePlatformConfig.h]: Reworked:
defined CONFIG_CHIP_THREAD_SSEDwhich is Boolean Kconfig macros is defined as true or not defined at all:
```
#ifndef CONFIG_CHIP_THREAD_SSED
#define CONFIG_CHIP_THREAD_SSED 0
#endif
```

File: src/platform/nrfconnect/CHIPDevicePlatformConfig.h [hunk 14]
```diff
@@ -217,7 +223,8 @@
...
+    CONFIG_CHIP_THREAD_SSED
```
[the remark this answers is not in the document]
Comment (s07641069) [reply src/platform/nrfconnect/CHIPDevicePlatformConfig.h]: Reworked:
defined CONFIG_CHIP_THREAD_SSED which is Boolean Kconfig macros is defined as true or not defined at all:
```
#ifndef CONFIG_CHIP_THREAD_SSED
#define CONFIG_CHIP_THREAD_SSED 0
#endif
```

File: src/platform/nrfconnect/CHIPDevicePlatformConfig.h [hunk 15]
```diff
@@ -246,7 +253,7 @@
 #endif // CONFIG_CHIP_EXTENDED_DISCOVERY
 
 #ifndef CHIP_DEVICE_CONFIG_ROTATING_DEVICE_ID_UNIQUE_ID_LENGTH
-#if CONFIG_CHIP_FACTORY_DATA
+#if defined(CONFIG_CHIP_FACTORY_DATA) && CONFIG_CHIP_FACTORY_DATA
```
[the remark this answers is not in the document]
Comment (s07641069) [reply src/platform/nrfconnect/CHIPDevicePlatformConfig.h]: Reworked:
defined CONFIG_CHIP_FACTORY_DATA which is Boolean Kconfig macros is defined as true or not defined at all:
```
#ifndef CONFIG_CHIP_FACTORY_DATA
#define CONFIG_CHIP_FACTORY_DATA 0
#endif
```

File: src/platform/nrfconnect/CHIPPlatformConfig.h [hunk 16]
```diff
@@ -70,7 +70,7 @@
 #define CHIP_CONFIG_MAX_FABRICS 5
 #endif
 
-#if CONFIG_CHIP_LOG_SIZE_OPTIMIZATION
+#if defined(CONFIG_CHIP_LOG_SIZE_OPTIMIZATION) && CONFIG_CHIP_LOG_SIZE_OPTIMIZATION
```
[the remark this answers is not in the document]
Comment (s07641069) [reply src/platform/nrfconnect/CHIPPlatformConfig.h]: Reworked:
defined CONFIG_CHIP_LOG_SIZE_OPTIMIZATION which is Boolean Kconfig macros is defined as true or not defined at all:
```
#ifndef CONFIG_CHIP_LOG_SIZE_OPTIMIZATION
#define CONFIG_CHIP_LOG_SIZE_OPTIMIZATION 0
#endif
```

File: src/platform/nrfconnect/ConnectivityManagerImpl.cpp [hunk 17]
```diff
@@ -51,7 +51,7 @@ CHIP_ERROR ConnectivityManagerImpl::_Init()
 #if CHIP_DEVICE_CONFIG_ENABLE_THREAD
     GenericConnectivityManagerImpl_Thread<ConnectivityManagerImpl>::_Init();
 #endif
-#if CHIP_DEVICE_CONFIG_ENABLE_WIFI
+#if defined(CHIP_DEVICE_CONFIG_ENABLE_WIFI) && CHIP_DEVICE_CONFIG_ENABLE_WIFI
```
[the remark this answers is not in the document]
Comment (s07641069) [reply src/platform/nrfconnect/ConnectivityManagerImpl.cpp]: Reworked:
defined CONFIG_WIFI_NRF700X which is Boolean Kconfig macros is defined as true or not defined at all:
```
#ifndef CONFIG_WIFI_NRF700X
#define CONFIG_WIFI_NRF700X 0
#endif
```

File: src/platform/nrfconnect/ConnectivityManagerImpl.h [hunk 18]
```diff
@@ -23,7 +23,7 @@
 #if INET_CONFIG_ENABLE_TCP_ENDPOINT
 #include <platform/internal/GenericConnectivityManagerImpl_TCP.h>
 #endif
-#if CHIP_DEVICE_CONFIG_ENABLE_CHIPOBLE
+#if defined(CHIP_DEVICE_CONFIG_ENABLE_CHIPOBLE) && CHIP_DEVICE_CONFIG_ENABLE_CHIPOBLE
```
[the remark this answers is not in the document]
Comment (s07641069) [reply src/platform/nrfconnect/ConnectivityManagerImpl.h]: Reworked:
defined CONFIG_BT which is Boolean Kconfig macros is defined as true or not defined at all:
```
#ifndef CONFIG_BT
#define CONFIG_BT 0
#endif
```

File: src/platform/nrfconnect/ExternalFlashManager.h [hunk 19]
```diff
@@ -35,7 +35,7 @@ class ExternalFlashManager
 
     virtual void DoAction(Action aAction)
     {
-#if CONFIG_PM_DEVICE && CONFIG_NORDIC_QSPI_NOR
+#if defined(CONFIG_PM_DEVICE) && defined(CONFIG_NORDIC_QSPI_NOR)
```
[the remark this answers is not in the document]
Comment (s07641069) [reply src/platform/nrfconnect/ExternalFlashManager.h:38]: Boolean Kconfig macros are either defined as true or not defined at all.

File: src/platform/nrfconnect/OTAImageProcessorImpl.cpp [hunk 20]
```diff
@@ -25,7 +25,7 @@
 #include <platform/CHIPDeviceLayer.h>
 #include <system/SystemError.h>
 
-#if CONFIG_CHIP_CERTIFICATION_DECLARATION_STORAGE
+#if defined(CONFIG_CHIP_CERTIFICATION_DECLARATION_STORAGE) && CONFIG_CHIP_CERTIFICATION_DECLARATION_STORAGE
```
[the remark this answers is not in the document]
Comment (s07641069) [reply src/platform/nrfconnect/OTAImageProcessorImpl.cpp]: Reworked:
defined CONFIG_CHIP_CERTIFICATION_DECLARATION_STORAGE which is Boolean Kconfig macros is defined as true or not defined at all:
```
#ifndef CONFIG_CHIP_CERTIFICATION_DECLARATION_STORAGE
#define CONFIG_CHIP_CERTIFICATION_DECLARATION_STORAGE 0
#endif // CONFIG_CHIP_CERTIFICATION_DECLARATION_STORAGE
```

File: src/platform/nrfconnect/wifi/WiFiManager.h (hunk 3 above)
Comment (s07641069) [reply src/platform/nrfconnect/wifi/WiFiManager.h]: Reworked:
defined CONFIG_CHIP_THREAD_SSED which is Boolean Kconfig macros is defined as true or not defined at all:
```
#ifndef CONFIG_CHIP_THREAD_SSED
#define CONFIG_CHIP_THREAD_SSED 0
#endif
```

File: src/system/SystemConfig.h (hunk 4 above)
Comment (s07641069) [reply src/system/SystemConfig.h:736]: Boolean Kconfig macros are either defined as true or not defined at all.

File: src/system/SystemConfig.h (hunk 5 above)
Comment (s07641069) [reply src/system/SystemConfig.h]: Boolean Kconfig macros are either defined as true or not defined at all.

File: src/system/SystemConfig.h [hunk 21]
```diff
@@ -764,6 +764,9 @@ struct LwIPEvent;
  *
  *  Defaults to enabled on Zephyr platforms that enable CONFIG_EVENTFD.
  */
+#ifndef CONFIG_EVENTFD
+#define CONFIG_EVENTFD 0
```
Comment (Damian-Nordic) [inline src/system/SystemConfig.h:768]: Can we check for `defined(CONFIG_EVENTFD)`, instead? All Kconfig booleans are either defined or not, so in this case setting default values is not necessary.

File: config/nrfconnect/chip-gn/args.gni [hunk 22]
```diff
@@ -20,8 +20,8 @@ chip_device_platform = "nrfconnect"
 
 chip_build_tests = false
 
-chip_project_config_include = ""
-chip_system_project_config_include = ""
+chip_project_config_include = "<CHIPProjectConfig.h>"
```
Comment (Damian-Nordic) [inline config/nrfconnect/chip-gn/args.gni:23]: Why is this necessary? These arguments are optional and we only set them if the corresponding Kconfig option is specified. What if a project does not provide the project config?

**Review by Damian-Nordic — CHANGES_REQUESTED** [review]
Can I prepare a patch to clear nRF config, which you could add to your PR? I don't quite like defining `CONFIG_XXX` macros as 0 or the need to add `CHIP_DEVICE_CONFIG_ENALE_SED` in the project config but I can help with resolving the original issues.

**Comment by s07641069** [comment]
Gave you access to my fork: https://github.com/Damian-Nordic
Please change NRF related configs.
> Can I prepare a patch to clear nRF config, which you could add to your PR? I don't quite like defining `CONFIG_XXX` macros as 0 or the need to add `CHIP_DEVICE_CONFIG_ENALE_SED` in the project config but I can help with resolving the original issues.

**Comment by Damian-Nordic** [comment]
@s07641069 OK, I pushed one commit. Let's see if I haven't broken anything :).

**Review by Damian-Nordic — APPROVED** [review]

**Review by bzbarsky-apple — APPROVED** [review]
